Ftr/dynamic annotation classification

bwbohl opened the pull request:
## Description, Context and related Issue
Add support for dynamic mei:taxonomy usage in mei annot

<!--- This project only accepts pull requests related to open issues. Please link to the issue here: -->
Refs https://github.com/Edirom/Edirom-Online-Backend/issues/81

## How Has This Been Tested?
* Klarinettenquintett dataset
* BAZ-GA dataset

## Types of changes
- New feature (non-breaking change which adds functionality)
- Documentation Update
- Improvement
- Refactoring

## Overview
- I have updated the inline documentation accordingly.
- I have performed a self-review of my code, according to the [style guide](https://github.com/Edirom/Edirom-Online/blob/develop/STYLE-GUIDE.md)
- I have read the [CONTRIBUTING](https://github.com/Edirom/Edirom-Online-Backend/blob/develop/CONTRIBUTING.md) document.

## TODOs

- [x] maintain API 1.0.0 compatibility
- [x] support old model of associating mei:annots with categories and priorities

bwbohl committed:
getAnnotations.xql: code style improvements

bwbohl committed:
getAnnotations.xql: code style improvements

bwbohl committed:
annotation.xqm: remove commented-out section

bwbohl committed:
annotation.xqm: remove commented-out section

bwbohl committed:
annotation.xqm: rework comment about $xsltBase

bwbohl committed:
annotation.xqm: rework comment about $xsltBase

bwbohl committed:
annotation.xqm: use eutil:getDoc

bwbohl committed:
annotation.xqm: use eutil:getDoc

bwbohl committed:
Create taxonomy.xqm

bwbohl committed:
Create taxonomy.xqm

bwbohl committed:
annotation.xqm: use new function  taxonomy:get-labels in getLocalizedLabel

bwbohl committed:
annotation.xqm: use new function  taxonomy:get-labels in getLocalizedLabel

bwbohl committed:
annotation.xqm: use switch in getLocalizeLabel

bwbohl committed:
annotation.xqm: use switch in getLocalizeLabel

bwbohl committed:
annotation.xqm: rename function

bwbohl committed:
annotation.xqm: rename function

bwbohl committed:
annotation.xqm: add some documentation

bwbohl committed:
annotation.xqm: add some documentation

bwbohl committed:
annotation.xqm: rename function

bwbohl committed:
annotation.xqm: rename function

bwbohl committed:
annotation.xqm: comment-out unused function

bwbohl committed:
annotation.xqm: comment-out unused function

bwbohl committed:
replace calls to annotation:category_getName with annotation:get-category-label-localized

bwbohl committed:
replace calls to annotation:category_getName with annotation:get-category-label-localized

bwbohl committed:
annotation.xqm: remove unsed function annotation:category_getName

bwbohl committed:
annotation.xqm: remove unsed function annotation:category_getName

bwbohl committed:
fix 2-arity calls to annotation:get-category-label-localized

bwbohl committed:
fix 2-arity calls to annotation:get-category-label-localized

bwbohl committed:
taxonomy.xqm: add XQDoc

bwbohl committed:
taxonomy.xqm: add XQDoc

bwbohl committed:
taxonomy.xqm: fix indentation

bwbohl committed:
taxonomy.xqm: fix indentation

bwbohl committed:
eutil.xqm: add generic is-empty function

bwbohl committed:
eutil.xqm: add generic is-empty function

bwbohl committed:
getAnnotations.xql: fetch individual annotations to variable

bwbohl committed:
getAnnotations.xql: fetch individual annotations to variable

bwbohl committed:
getAnnotations.xql: new entry field in return, listing annotation field names

bwbohl committed:
getAnnotations.xql: new entry field in return, listing annotation field names

bwbohl committed:
getAnnotations.xql: new return entry listing fields empty in all annotations

bwbohl committed:
getAnnotations.xql: new return entry listing fields empty in all annotations

bwbohl committed:
annotation:toJSON : move definition of return map to variable

bwbohl committed:
annotation:toJSON : move definition of return map to variable

bwbohl committed:
taxonomy.xqm: add new function to get localized label as string

bwbohl committed:
taxonomy.xqm: add new function to get localized label as string

bwbohl committed:
annotation.xqm: use taxonomy:get-label-localized-as-string

bwbohl committed:
annotation.xqm: use taxonomy:get-label-localized-as-string

bwbohl committed:
annotation.xqm: new function  annotation get-class-idrefs-as-sequence

bwbohl committed:
annotation.xqm: new function  annotation get-class-idrefs-as-sequence

bwbohl committed:
annotation.xqm: annotation:get-class-labels-as-sequence

bwbohl committed:
annotation.xqm: annotation:get-class-labels-as-sequence

bwbohl committed:
taxonomy.xqm: new function taxonomy:get-root-identifying-string

bwbohl committed:
taxonomy.xqm: new function taxonomy:get-root-identifying-string

bwbohl committed:
taxonomy.xqm: use taxonomy:get-root-identifying-string

bwbohl committed:
taxonomy.xqm: use taxonomy:get-root-identifying-string

bwbohl committed:
remove whitespace

bwbohl committed:
remove whitespace

bwbohl committed:
taxonomy.xqm: add some inline documentation

bwbohl committed:
taxonomy.xqm: add some inline documentation

bwbohl committed:
annotation:toJSON use annotation:get-class-idrefs-as-sequence for gettin annotation’s classes

bwbohl committed:
annotation:toJSON use annotation:get-class-idrefs-as-sequence for gettin annotation’s classes

bwbohl committed:
annotation:toJSON create map entries for each taxonomy listing referenced categories

bwbohl committed:
annotation:toJSON create map entries for each taxonomy listing referenced categories

bwbohl committed:
annotation:toJSON remove categories and priority from $baseMap

bwbohl committed:
annotation:toJSON remove categories and priority from $baseMap

bwbohl committed:
annotation:toJSON add some inline documentation

bwbohl committed:
annotation:toJSON add some inline documentation

bwbohl committed:
edirom-lang-de: translate untranslated value

bwbohl committed:
edirom-lang-de: translate untranslated value

bwbohl committed:
lang-files: formatting, create separat AnnotView block

bwbohl committed:
lang-files: formatting, create separat AnnotView block

bwbohl committed:
lang-files: harmonize AnnotationView values

bwbohl committed:
lang-files: harmonize AnnotationView values

bwbohl committed:
remove old TODO

bwbohl committed:
remove old TODO

bwbohl committed:
annotation:toJSON – string-join return value for taxonomies

bwbohl committed:
annotation:toJSON – string-join return value for taxonomies

bwbohl committed:
locale: fix the keys for AnnotationView categories and priorities

bwbohl committed:
locale: fix the keys for AnnotationView categories and priorities

bwbohl commented:
Please be aware that this has to be tested in combination with: https://github.com/Edirom/Edirom-Online-Frontend/pull/87

daniel-jettka commented:
Hi @bwbohl Thanks. Can you give some more detail about the feature? Some description about changes in data and some screenshots would be great.

@Edirom/edirom-online-developer Is anyone able to review this in the next two days until the test phase begins? :-)

bwbohl commented:
Some background:

The BAZ-GA uses different classification systems than ediromCategory and ediromPriority, namely a bazga.category and a bazga.class, which are defined in two mei:taxonomies. This led to an empty "priority"-column in the annotation list or single views and an over-populated "category"-column. Moreover, we wanted our class and category in separate columns.

Instead of creating a hard-coded fork of the frontend just for the BAZ-GA, we decided to push the already data-driven (mei:taxonomies were already evaluated for the category and priority columns) processing of the Edirom even further by:

1. Not generally providing priority and category data fields from the backend, but get all the taxonomies used in a list of annotations and create a column for each taxonomy
2. In the frontend, by default, hide columns that are empty in all annotations

This, of course, results in an "unstable" annotation model – so the frontend annotation model had to be skipped and the store and table built according to the data returned.

For all editions that use EdiromCategory and EdiromPriority, everything should look as before, but for those using other custom taxonomies, these should have their own columns now.

bwbohl commented:
As stated yesterday at the [Community Meeting](https://github.com/Edirom/Edirom-Online/wiki/2025-10-01-Edirom-Community-Meeting), I realised that the single annotation view is still missing the new concept. I will make this pull request a draft until I’ve added the necessary changes.

bwbohl commented:
I have updated the single view and, while doing so, reactivated the `sourceLabels` div in the metaBox because I don’t like commented out code ;-) If there is a need for individual fields being hidden we would have to implement this ina different way.

Observation:
For the metaBox, the Backend returns HTML – I’m not totally against it, but I’d rather put something like that into the frontend. But there might be other places where the backend behaves similarly. I think this would need a more general discussion and conception.

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

bwbohl commented:
@daniel-jettka I reviewed the current state: apparently, there’s only one thing left so far that I have to fix, being getAnnotationInfos.xql, which still only returns priorities and categories. This is why newly created taxonomy trees don’t show up in the Annotation menu as you cescribed in https://github.com/Edirom/Edirom-Online-Frontend/pull/87#pullrequestreview-3308994097

Will fix until tomorrow’s Community Meeting ;-)

bwbohl committed:
getAnnotationInfos.xql: use taxonomy module to return dynamic categories’ array

bwbohl committed:
getAnnotationInfos.xql: use taxonomy module to return dynamic categories’ array

bwbohl committed:
taxonomy.xqm: fix sequence order in taxonomy:get-root-identifying-string

bwbohl committed:
taxonomy.xqm: fix sequence order in taxonomy:get-root-identifying-string

bwbohl committed:
taxonomy.xqm: introduce get-parent-taxonomy-identifying-string function

bwbohl committed:
taxonomy.xqm: introduce get-parent-taxonomy-identifying-string function

bwbohl committed:
annotation:toJSON – use new taxonomy:get-parent-taxonomy-identifying-string function

bwbohl committed:
annotation:toJSON – use new taxonomy:get-parent-taxonomy-identifying-string function

bwbohl committed:
getAnnotationMeta: remodel to support dynamic taxonomies

bwbohl committed:
getAnnotationMeta: remodel to support dynamic taxonomies

bwbohl committed:
fix: getAnnotations.xql to consider all annotationFields not just for the first annotation

bwbohl committed:
fix: getAnnotations.xql to consider all annotationFields not just for the first annotation

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

bwbohl commented on data/xqm/annotation.xqm:
```diff
@@ -80,26 +99,52 @@ declare function annotation:toJSON($anno as element(), $edition as xs:string) as
+
+    (: create a map with all static information about the annotation :)
+    let $baseMap := map {
+        'id': $id,
+        'title': normalize-space($title),
+        'pos': string($count),
+        'sigla': string-join($sigla,', ')
+    }
+
+    (: create a map with keys for each taxonomy used for this annotation and the corresponding class labels :)
+    let $taxonomiesMap := map:merge(
+        for $usedTaxonomy in $classes-elements[ancestor::mei:taxonomy]
+        let $taxonomyIdentifier := taxonomy:get-parent-taxonomy-identifying-string( $usedTaxonomy )
+        return
+            map:entry(
+                $taxonomyIdentifier,
+                string-join (
+                    (
+                        for $classElement in $classes-elements[self::mei:category]
+                        where taxonomy:get-parent-taxonomy-identifying-string( $classElement ) = $taxonomyIdentifier
+                        return taxonomy:get-label-localized-as-string($classElement)
+                    ),
+                    ', '
+                )
+            )
+    )
+
     return
-        map {
-            'id': $id,
-            'title': normalize-space($title),
-            'categories': $cats,
-            'priority': $prio,
-            'pos': string($count),
-            'sigla': string-join($sigla,', ')
-        }
+        map:merge((
+            $baseMap,
+            $taxonomiesMap
+        ))
```
@Edirom/edirom-online-developer With these changes, I guess most of you would consider this a breaking change to the API, right?

daniel-jettka replied:
Same like other one, it keeps the change as backwards-compatible as possible

bwbohl commented on data/xql/getAnnotationInfos.xql:
```diff
@@ -87,33 +37,38 @@ let $mei := doc($uri)
+                return
+                    map {
+                        'id': xs:string($id),
+                        'name': taxonomy:get-label-localized-as-string($e)
+                    }
+            }
         }
-
-let $prios :=
-    for $priority in local:getDistinctPriorities($annots)
-    let $name := annotation:getPriorityLabel(($editionCollection/id($priority)[mei:label or mei:name])[1])
-    order by $name
-    return
-        map {
-            'id': $priority,
-            'name': $name
-        }
-
-let $map :=
-   map {
-       'categories': array {$categories},
-       'priorities': array {$prios},
-       'count': count($annots)
-   }
+}
 
 return
-    $map
+    map {
+        'count': count($annots),
+        'taxonomies': $taxonomiesArray
+    }
```
@Edirom/edirom-online-developer With these changes, I guess most of you would consider this a breaking change to the API, right?

daniel-jettka replied:
Now the change keeps it as backwards-compatible as possible

bwbohl commented on data/xql/getAnnotations.xql:
```diff
@@ -7,38 +7,65 @@ xquery version "3.1";
     }
-
-return
-    $map
```
@Edirom/edirom-online-developer With these changes, I guess most of you would consider this a breaking change to the API, right?

bwbohl replied:
'annotations' now reinstated

bwbohl commented:
@Edirom/edirom-online-developer I’d really like to hear your thoughts on my comments above, because, as I gather from other pull requests, changes that affect the API are being postponed to a later release, right?

peterstadler commented:
Right, I'm always in favor of keeping the current endpoints stable and only implement breaking changes in the API v2.

But wouldn't it be possible here to simply add your proposed new properties to the response objects (without removing the old properties)?

hizclick commented:
I agree with @peterstadler, we can postpone this PR to the next release.

bwbohl commented:
> Right, I'm always in favor of keeping the current endpoints stable and only implement breaking changes in the API v2.
> 
> But wouldn't it be possible here to simply add your proposed new properties to the response objects (without removing the old properties)?

If that’s ok, then I’ll try to add the new stuff as an additional return content. The frontend changes are qute drastic though and might just ignore the old stuff…

bwbohl committed:
annotation:get-referenced-category-element new function

bwbohl committed:
annotation:get-referenced-category-element new function

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

hizclick commented:
@bwbohl What do you think if we postpone this PR to the next release?

hizclick commented:
> @bwbohl What do you think if we postpone this PR to the next release?

Since the deadline for the code freeze has been extended, I have changed my recommendation and no longer suggest postponing this to the next release.

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

bwbohl committed:
fix wrong characters

bwbohl committed:
fix wrong characters

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

bwbohl committed:
taxonomy:get-label-localized-as-string: use new function eutil:getSetLang

bwbohl committed:
taxonomy:get-label-localized-as-string: use new function eutil:getSetLang

bwbohl committed:
taxonomy:get-parent-taxonomy-identifying-string : fix precendence of declarations

bwbohl committed:
taxonomy:get-parent-taxonomy-identifying-string : fix precendence of declarations

bwbohl committed:
new function annotation:get-referenced-categories-as-taxonomy-array

bwbohl committed:
new function annotation:get-referenced-categories-as-taxonomy-array

bwbohl committed:
getAnnotationInfos.xql: filter mei:annots to be of type editorialComment

bwbohl committed:
getAnnotationInfos.xql: filter mei:annots to be of type editorialComment

bwbohl committed:
Align how AnnotationInfo AnnotationMeta get annotation categories

use new function annotation:get-referenced-categories-as-taxonomy-array

bwbohl committed:
Align how AnnotationInfo AnnotationMeta get annotation categories

use new function annotation:get-referenced-categories-as-taxonomy-array

bwbohl committed:
remodel function annotation:get-referenced-category-elements

bwbohl committed:
remodel function annotation:get-referenced-category-elements

bwbohl committed:
annotation:annotationsToJSON cleanup

bwbohl committed:
annotation:annotationsToJSON cleanup

bwbohl committed:
taxonomy:get-label-localized-as-string accept mei:taxonomy

bwbohl committed:
taxonomy:get-label-localized-as-string accept mei:taxonomy

bwbohl committed:
taxonomy:get-label-localized-as-string add xml:id as fallback return value

bwbohl committed:
taxonomy:get-label-localized-as-string add xml:id as fallback return value

bwbohl committed:
taxonomy:get-label-localized-as-string add description

bwbohl committed:
taxonomy:get-label-localized-as-string add description

bwbohl committed:
annotation:get-referenced-categories-as-taxonomy-array use taxonomy function to return label

bwbohl committed:
annotation:get-referenced-categories-as-taxonomy-array use taxonomy function to return label

bwbohl committed:
taxonomy:get-label-localized-as-string reworked to allow empty returns

bwbohl committed:
taxonomy:get-label-localized-as-string reworked to allow empty returns

bwbohl committed:
RE-ADD LEGACY SUPPORT for API v1.0.0 conformance

Add function declarations to extract distinct categories and
priorities from annotations, supporting both legacy pre-MEI 4 and
modern MEI 4+ models. Include these fields in getAnnotationInfos
response.

Conditionally populate legacyFields in getAnnotations only when
taxonomy-derived fields are present in the data.

Add edition module import and deprecation handling in
getAnnotationMeta for legacy annotation fields.

feat: add annotation_hide_legacy_fields preference

Add new preference to hide legacy annotation fields
(categories, priority) in the annotation grid on
initial render. This prevents duplicate columns for
editions using ediromCategory/ediromPriority
taxonomies. Defaults to true but can be disabled per
edition.

bwbohl committed:
RE-ADD LEGACY SUPPORT for API v1.0.0 conformance

Add function declarations to extract distinct categories and
priorities from annotations, supporting both legacy pre-MEI 4 and
modern MEI 4+ models. Include these fields in getAnnotationInfos
response.

Conditionally populate legacyFields in getAnnotations only when
taxonomy-derived fields are present in the data.

Add edition module import and deprecation handling in
getAnnotationMeta for legacy annotation fields.

feat: add annotation_hide_legacy_fields preference

Add new preference to hide legacy annotation fields
(categories, priority) in the annotation grid on
initial render. This prevents duplicate columns for
editions using ediromCategory/ediromPriority
taxonomies. Defaults to true but can be disabled per
edition.

bwbohl committed:
refactor: standardize annotation view keys

Rename annotation view language keys to use consistent
edirom prefix naming convention. Update
AnnotationView_Categories to AnnotationView_ediromCategories
and AnnotationView_Priority to
AnnotationView_ediromPriority across all locale files and
query logic. This ensures uniform key naming for all
edirom-specific annotation view labels in German and
English locales.

Squashed commit of the following:

commit fc0388694e283ec83d748d2e812a25ed1f9f90a1
Author: bwbohl <b.w.bohl@gmail.com>
Date:   Thu May 28 05:25:59 2026 +0200

    Mo-Stashed changes

bwbohl committed:
refactor: standardize annotation view keys

Rename annotation view language keys to use consistent
edirom prefix naming convention. Update
AnnotationView_Categories to AnnotationView_ediromCategories
and AnnotationView_Priority to
AnnotationView_ediromPriority across all locale files and
query logic. This ensures uniform key naming for all
edirom-specific annotation view labels in German and
English locales.

bwbohl committed:
allow empty parameter in annotation:get-referenced-categories-as-taxonomy-array

Change annotation parameter from required to optional by updating
the cardinality from + to *. Wrap the main logic in a try-catch
block to gracefully handle errors by returning an empty array.

bwbohl committed:
getAnnotationsOnPage: return taxonomyClasses as annotation property

Extract taxonomy classes from annotation element classes
and include them in the returned annotation object.

bwbohl committed:
feat: enhance taxonomy label resolution with fallback chain

Extend the label resolution logic to provide comprehensive
fallback options when no matching language is found. When the
languages array is empty, return labels for all available
localizations including an 'und' (undefined) entry.

The fallback chain now resolves labels in this order:
1. child mei:label without @xml:lang (mei:category only)
2. @label attribute
3. first mei:label child element (mei:category only)
4. element's @xml:id
5. parent taxonomy's identifying string

This ensures that taxonomy and category elements are always
resolvable with meaningful identifiers, even when explicit
labels are unavailable.

bwbohl committed:
feat: add function to resolve cross-file references

Add eutil:get-referenced-element function to resolve single
reference tokens to their target elements by id. The function
handles both fragment-only tokens and cross-file references,
resolving them within the referencing document's context or
loading external documents as needed.

bwbohl committed:
refactor: simplify cross-file element resolution

Replace manual document resolution logic with eutil:get-referenced-element
utility function throughout annotation processing. This eliminates
redundant code for handling both local and cross-file references.

Simplify category and class label retrieval by using dedicated helper
functions. Remove unnecessary intermediate variables and streamline
tokenization logic for @class and @ptr attributes.

bwbohl committed:
fix: improve annotation label resolution logic

Refactor the annotation metadata label resolution to mirror
frontend behavior. Now prioritizes explicit labels over locale
file lookups, using the identifier as a fallback. This ensures
consistent label handling between backend and frontend while
maintaining proper internationalization support.

bwbohl committed:
feat: handle plural labels for multi-item annotations

Add logic to distinguish between singular and plural
label strings based on item count. When an annotation
has multiple items, use the '_multiple' variant of the
label key to support proper internationalization of
plural forms.

bwbohl committed:
fix: improve annotation URI matching logic

Replace simple regex matching with tokenized URI comparison
to properly handle annotation plists containing multiple
space-separated values. This ensures annotations are
correctly matched even when their plist attribute contains
multiple URIs or URI fragments and avoids false positive matches.

bwbohl committed:
ftr: add tests for taxonomy module

THe taxnomy module introduced with this ull request also needs a test
module. This is a first go (not covering every detail I guess)

bwbohl committed:
ftr: add annotation.xqm tests

This pull request is all about annotations, so adding a first draft for
a test module seemed plausible.

bwbohl committed:
ftr: add support for plural forms of taxonomy labels

bwbohl committed:
Remove duplicate import of eutil module

Removed duplicate import of eutil module.

Signed-off-by: Benjamin W. Bohl <b.w.bohl@gmail.com>

daniel-jettka reviewed (approved):
Tested together with https://github.com/Edirom/Edirom-Online-Frontend/pull/87

Also see approval of frontend PR: https://github.com/Edirom/Edirom-Online-Frontend/pull/87#pullrequestreview-4624260191
